### PR TITLE
feat(lineage): RFC for column level lineage 

### PR DESCRIPTION
**Pull request by liangjun-jiang**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by mars-lan** [comment]
Rendered doc: https://github.com/liangjun-jiang/datahub/blob/rfc-column-level-lineage/docs/rfc/column-level-lineage.md

File: docs/rfc/column-level-lineage.md [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+                            "fields": [
+                                {
+                                    "sourceField": "foo131",
+                                    "targetField" : "bar131"
+                                },
+                                {
+                                    "sourceField": "foo132",
+                                    "targetField" : "bar132"
+                                }
+                            ]
```
Comment (nagarjunakanamarlapudi) [inline docs/rfc/column-level-lineage.md:73]: We might have to start thinking field as a first class entity. And, the relations to the source fields can come from multiple datasets. Imagine, two fields from two different datasets can get transformed to form a destination field. 

```
record DatasetFieldMapping includes BaseFieldMapping{
 /**
  * Source fields from which the fine grained lineage is derived
  */
 sourceFields: union[DatasetFieldUrn]

 /**
  * Destination field which is derived from source fields
  */
 destinationField: DatasetFieldUrn
```

File: docs/rfc/column-level-lineage.md [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+ firstly, we need to establish a relationship between `dataset` and its `fields`. Let's call it `hasField` for right now.
+
+In the `neo4j`, the expected graph representation would look like:
+> disclaimer: I am not a neo4j expert, or actually know too much about it. 
+
+![column-lineage-diagram](https://user-images.githubusercontent.com/356347/87818179-ca396a80-c82f-11ea-95d6-60eb1601c994.png)
+
+To have this new relationship, a few files would be changed. We can use this [Onboarding to GMA Graph - Adding a new relationship type](https://github.com/linkedin/datahub/blob/master/docs/demo/graph-onboarding.md) as the reference for the implementation detail.
+
+#### 2. Establish a relationship between a field of a source dataset and a field of target dataset.
```
Comment (nagarjunakanamarlapudi) [inline docs/rfc/column-level-lineage.md:521]: Referring to my previous comment .. the source fields can be more than one.

File: docs/rfc/column-level-lineage.md [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+6. bar 132 of c_bar_13 is derived by foo132 and 2_foo132
+
+
+
+![Screen Shot 2020-08-04 at 11 42 54 AM](https://user-images.githubusercontent.com/356347/89320841-a8215400-d647-11ea-882b-8c9dea4bdbf6.png)
+
+
+
+## Drawbacks
+Because `SchemaField` is not a first-class entity in Datahub, the current [Neo4jGraphWriterDao](https://github.com/linkedin/datahub/blob/master/metadata-dao-impl/neo4j-dao/src/main/java/com/linkedin/metadata/dao/internal/Neo4jGraphWriterDAO.java) only supports the first class entity by enforcing [validation rule] implemented by [RelationshipValidator](https://github.com/linkedin/datahub/blob/master/metadata-validators/src/main/java/com/linkedin/metadata/validator/RelationshipValidator.java)
```
Comment (nagarjunakanamarlapudi) [inline docs/rfc/column-level-lineage.md:704]: Thanks for your work with work arounds to get this working. 

Search and Discovery of fields and its associated metadata is also one of the common asks I believe. We might also want to think more inlines of making datasetfield as an entity in itself.

File: docs/rfc/column-level-lineage.md (hunk 1 above)
Comment (liangjun-jiang) [reply docs/rfc/column-level-lineage.md:73]: Right, source fields can be more than one. I define the 
```
record FineGrainUpstream {
  auditStamp: AuditStamp
  dataset: DatasetUrn
  fields: array[FineGrainFieldUpstream]
}
```
The `fields` is an array of the FieldMapping, which, in my opinion, could work with current neo4j implementation, and generate `n:1` relationship between a group of `sourceField` and `a destination`. 
The reason to implement in this way is to get around of the problem that neo4j only recognizes the first class entity. 

For sure, if we have made datasetField as a new entity, the problem above might not be existed. It however will create different problem on the neo4j graph creation.

Comment (liangjun-jiang) [reply docs/rfc/column-level-lineage.md:73]: I have thought about the DatasetField as the first class entity . As I remember, `Schema` being a first class entity is on the [road map](https://github.com/linkedin/datahub/blob/master/docs/roadmap.md#schemas-as-an-entity). To me, we should onboard `schema` before onboarding `datasetField`.

**Pull request closed without merging**